coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeIncrementalTransformer.java

```diff
@@ -0,0 +1,109 @@
+        RelOptTable incrementalTable =
+            RelOptTableImpl.create(originalTable.getRelOptSchema(), originalTable.getRowType(), incrementalNames, null);
+        return LogicalTableScan.create(scan.getCluster(), incrementalTable);
+      }
+
+      @Override
+      public RelNode visit(LogicalJoin join) {
+        RelNode left = join.getLeft();
+        RelNode right = join.getRight();
+        RelNode incrementalLeft = convertRelIncremental(left);
+        RelNode incrementalRight = convertRelIncremental(right);
+
+        RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+
+        LogicalProject p1 = createProjectOverJoin(join, left, incrementalRight, rexBuilder);
+        LogicalProject p2 = createProjectOverJoin(join, incrementalLeft, right, rexBuilder);
+        LogicalProject p3 = createProjectOverJoin(join, incrementalLeft, incrementalRight, rexBuilder);
+
+        LogicalUnion unionAllJoins =
+            LogicalUnion.create(Arrays.asList(LogicalUnion.create(Arrays.asList(p1, p2), true), p3), true);
+        return unionAllJoins;
+      }
+
+      @Override
+      public RelNode visit(LogicalFilter filter) {
+        RelNode transformedChild = convertRelIncremental(filter.getInput());
+        return LogicalFilter.create(transformedChild, filter.getCondition());
+      }
+
+      @Override
+      public RelNode visit(LogicalProject project) {
+        RelNode transformedChild = convertRelIncremental(project.getInput());
+        return LogicalProject.create(transformedChild, project.getProjects(), project.getRowType());
+      }
+
+      @Override
+      public RelNode visit(LogicalUnion union) {
+        List<RelNode> children = union.getInputs();
+        List<RelNode> transformedChildren =
+            children.stream().map(child -> convertRelIncremental(child)).collect(Collectors.toList());
+        return LogicalUnion.create(transformedChildren, union.all);
+      }
+
+      @Override
+      public RelNode visit(LogicalAggregate aggregate) {
+        RelNode transformedChild = convertRelIncremental(aggregate.getInput());
+        return LogicalAggregate.create(transformedChild, aggregate.getGroupSet(), aggregate.getGroupSets(),
+            aggregate.getAggCallList());
+      }
+    };
+    return originalNode.accept(converter);
+  }
+
+  private static LogicalProject createProjectOverJoin(LogicalJoin join, RelNode left, RelNode right,
+      RexBuilder rexBuilder) {
+    LogicalJoin incrementalJoin =
+        LogicalJoin.create(left, right, join.getCondition(), join.getVariablesSet(), join.getJoinType());
+    ArrayList<RexNode> projects = new ArrayList<>();
+    ArrayList<String> names = new ArrayList<>();
+    IntStream.range(0, incrementalJoin.getRowType().getFieldList().size()).forEach(i -> {
+      projects.add(rexBuilder.makeInputRef(incrementalJoin, i));
+      names.add(incrementalJoin.getRowType().getFieldNames().get(i));
+    });
+    return LogicalProject.create(incrementalJoin, projects, names);
```

Comment:
I recall we discussed creating the projection was not necessary?

Reply:
Looks like returning a LogicalJoin works correctly on the SQL level (passes the tests) but generates a RelNode representation that is different than what Coral would generate going from the same SQL -> RelNode. I think it might be better to create the projection to maintain consistency on the RelNode level.